File: nodejs/src/bordas/pessoa-borda.js
const MsgErro = require('../entidades/msg-erro');
const Pessoa = require('../entidades/db/pessoa');
const Carro = require('../entidades/db/carro');
const { Op } = require("sequelize");

module.exports = (app) => {
  const base = '/pessoa';

  app.post(base+'', async (req, res) => {
    if( !req.body.nome ) return res.status(400).send(new MsgErro(0,"Campo 'nome' (string) obrigatório!"));
    const pessoa = await Pessoa.create({ nome: req.body.nome });
    return res.status(200).json(pessoa);
  });

  app.get(base+'/todos', async (req, res) => {
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    const lista = await Pessoa.findAll({ include: Carro, offset: page*size, limit: size });
    return res.status(200).json(lista);
  });

  app.get(base+'', async (req, res) => {
    const page = parseInt(req.query.page) || 0;
    const size = parseInt(req.query.size) || 10;
    const id = parseInt(req.query.id) || 0;
    const nome = req.query.nome || '';
    const where = {};
    if( id ) where.id = id;
    if( nome.trim() ) where.nome = { [Op.like]: `%${nome.trim()}%` };
    const lista = await Pessoa.findAll({ include: Carro, offset: page*size, limit: size, where:where });
    return res.status(200).json(lista);
  });

};

File: golang/src/bordas/pessoa-borda.go
package bordas

import (
	"docker-memory/golang/src/config"
	"docker-memory/golang/src/entidades"
	"docker-memory/golang/src/entidades/db"
	"log"
	"strings"

	"github.com/gin-gonic/gin"
)

func InitPessoa(r *gin.Engine) {
	base := "/pessoa"

	r.POST(base+"/", func(c *gin.Context) {
		var params db.Pessoa
		c.ShouldBindJSON(&params)
		if params.Nome == "" {
			c.JSON(400, entidades.MsgErro{Msg: "Campo 'nome' (string) obrigatório!"})
			return
		}
		res := config.DB.Create(&params)
		if res.Error != nil {
			log.Println(res.Error.Error())
			c.JSON(500, entidades.MsgErro{Msg: "Erro ao guardar dados!"})
			return
		}
		c.JSON(200, params)
	})

	r.GET(base+"/todos", func(c *gin.Context) {
		var params entidades.Paging
		c.ShouldBindQuery(&params)
		params.CheckDefaults()

		var pessoas []db.Pessoa
		config.DB.Limit(params.Size).Offset(params.Page * params.Size).Preload("Carros").Find(&pessoas)
		c.JSON(200, pessoas)
	})

	r.GET(base+"/", func(c *gin.Context) {
		var params entidades.Paging
		c.ShouldBindQuery(&params)
		params.CheckDefaults()
		var pessoaParams db.Pessoa
		c.ShouldBindQuery(&pessoaParams)

		sql := config.DB.Limit(params.Size).Offset(params.Page * params.Size).Preload("Carros")
		if pessoaParams.Id > 0 {
			sql = sql.Where("id = ?", pessoaParams.Id)
		}
		pessoaParams.Nome = strings.TrimSpace(pessoaParams.Nome)
		if pessoaParams.Nome != "" {
			sql = sql.Where("nome LiKE ?", "%"+pessoaParams.Nome+"%")
		}

		var pessoas []db.Pessoa
		sql.Find(&pessoas)
		c.JSON(200, pessoas)
	})

}
